serve/mlc_serve/model/paged_cache_model.py

```diff
@@ -81,60 +186,100 @@ def __init__(
         for id, size in zip(sequence_ids, target_sizes):
             num_needed_block = math.ceil(size / self.block_size)
 
             if self.block_sliding_window:
                 num_needed_block = min(num_needed_block, self.block_sliding_window)
 
-            if id in self.kv_cache.block_tables and size == 0:
-                self.free_blocks.extend(self.kv_cache.block_tables[id])
-                del self.kv_cache.block_tables[id]
-                del self.kv_cache.slot_mappings[id]
+            if size == 0:
+                if id in self.kv_cache.prompt_block_tables:
+                    self.free_blocks.extend(self.kv_cache.prompt_block_tables[id])
+                    del self.kv_cache.prompt_block_tables[id]
+                elif id in self.kv_cache.decode_block_tables:
+                    self.free_blocks.extend(
+                        self.kv_cache.decode_block_tables[id].decode_blocks
+                    )
+                    del self.kv_cache.decode_block_tables[id]
+
+                if id in self.kv_cache.slot_mappings:
+                    del self.kv_cache.slot_mappings[id]
 
-            elif id in self.kv_cache.block_tables:
-                # Decoding
-                if len(self.kv_cache.block_tables[id]) < num_needed_block:
+            elif id in self.kv_cache.decode_block_tables:
+                decode_block_table = self.kv_cache.decode_block_tables[id]
+
+                if len(decode_block_table) < num_needed_block:
                     # Need to allocate a new block for this request
-                    assert len(self.kv_cache.block_tables[id]) + 1 == num_needed_block
-                    self.kv_cache.block_tables[id].append(self.free_blocks.pop())
+                    assert len(decode_block_table) + 1 == num_needed_block
+                    decode_block_table.append(self.free_blocks.pop())
 
                 pos = size - 1
-                block_number = self.kv_cache.block_tables[id][-1]
 
-                if self.block_sliding_window:
-                    block_number = self.kv_cache.block_tables[id][
-                        (pos // self.block_size) % self.block_sliding_window
-                    ]
+                def get_block_circular_index(token_pos):
+                    assert self.block_sliding_window
+                    return (token_pos // self.block_size) % self.block_sliding_window
+
+                if (
+                    decode_block_table.prompt_shared
+                    and self.sliding_window
+                    and size >= self.sliding_window
+                ):
+                    # Parallel sampling + SWA case
+                    if decode_block_table.prompt_cursor == get_block_circular_index(
+                        pos
+                    ):
+                        # This sequence is trying to overwrite a prompt block shared with other sequences.
+
+                        # TODO(masahi): The engine should take into account this additional
```

Comment:
Comment is not clear. The verification and handling seems is correct. What additionally engine should take into account and in which case?

Reply:
Thinking about it more, we probably don't need to worry about this since the engine conservatively assumes that each decode step can allocate one block for all sequences (correct me if I'm wrong about this). Whenever we hit this code path, we don't do allocation at L226. So we only allocate up to one block per decode step in all code paths.